tokio/src/sync/mpsc/block.rs

```diff
@@ -71,28 +78,54 @@ pub(crate) fn offset(slot_index: usize) -> usize {
     SLOT_MASK & slot_index
 }
 
+generate_addr_of_methods! {
+    impl<T> Block<T> {
+        unsafe fn addr_of_header(self: NonNull<Self>) -> NonNull<BlockHeader<T>> {
+            &self.header
+        }
+
+        unsafe fn addr_of_values(self: NonNull<Self>) -> NonNull<Values<T>> {
+            &self.values
+        }
+    }
+}
+
 impl<T> Block<T> {
-    pub(crate) fn new(start_index: usize) -> Block<T> {
-        Block {
-            // The absolute index in the channel of the first slot in the block.
-            start_index,
+    pub(crate) fn new(start_index: usize) -> Box<Block<T>> {
+        unsafe {
+            // Allocate the block on the heap.
+            let block = std::alloc::alloc(Layout::new::<Block<T>>()) as *mut Block<T>;
+            let block = match NonNull::new(block) {
+                Some(block) => block,
+                None => std::alloc::handle_alloc_error(Layout::new::<Block<T>>()),
+            };
+
+            // Write the header to the block.
+            Block::addr_of_header(block).as_ptr().write(BlockHeader {
+                // The absolute index in the channel of the first slot in the block.
+                start_index,
+
+                // Pointer to the next block in the linked list.
+                next: AtomicPtr::new(ptr::null_mut()),
 
-            // Pointer to the next block in the linked list.
-            next: AtomicPtr::new(ptr::null_mut()),
+                ready_slots: AtomicUsize::new(0),
 
-            ready_slots: AtomicUsize::new(0),
+                observed_tail_position: UnsafeCell::new(0),
+            });
 
-            observed_tail_position: UnsafeCell::new(0),
+            // Initialize the values array.
+            let values = Block::addr_of_values(block);
+            Values::<T>::initialize(&mut *values.cast().as_ptr());
```

Comment:
This can be simplified a bit.
```suggestion
            // Initialize the values array.
            Values::<T>::initialize(Block::addr_of_values(block).cast().as_mut());
```

Reply:
I mean, really, this method could just a raw pointer since it immediately converts it back into one.